dnscrypt-proxy/config.go

```diff
@@ -712,8 +712,12 @@ func (config *Config) loadSource(proxy *Proxy, requiredProps stamps.ServerInform
 	}
 	source, err := NewSource(cfgSourceName, proxy.xTransport, cfgSource.URLs, cfgSource.MinisignKeyStr, cfgSource.CacheFile, cfgSource.FormatStr, time.Duration(cfgSource.RefreshDelay)*time.Hour)
 	if err != nil {
-		dlog.Criticalf("Unable to retrieve source [%s]: [%s]", cfgSourceName, err)
-		return err
+		if len(source.in) > 0 { // cache loaded, ignore download failure
+			dlog.Infof("Downloading [%s] failed: %v, use cache file to startup", source.name, err)
+		} else {
+			dlog.Criticalf("Unable to retrieve source [%s]: [%s]", cfgSourceName, err)
+			return err
+		}
```

Comment:
If you swap these 2 around so the `if` checks the fatal condition, then the non-fatal path doesn't need to be in an `else` block. This would also match the style used later in the file to check the next "maybe-fatal" condition.

Reply:
Yeah, you are right. It gives a tighter code and matches the code style.
I took your advice, and appreciate your patience go guide me, even though you got the right pointer quicker. This really helps beginners who willing to contribute to get involved. Thanks!